packages/next/client/link.tsx

```diff
@@ -94,8 +94,8 @@ function linkClicked(
   e.preventDefault()
 
   //  avoid scroll for urls with anchor refs
-  if (scroll == null) {
-    scroll = as.indexOf('#') < 0
+  if (scroll == null && as.indexOf('#') >= 0) {
```

Comment:
This change reverts scroll always being defined, as this interfered with the combination of the default scroll behavior for shallow routing and setting the scroll option to false.